File: api/src/main/java/net/blueshell/api/base/JpaListener.java
package net.blueshell.api.base;

import jakarta.persistence.*;
import lombok.Setter;
import net.blueshell.api.common.event.jpa.*;
import org.springframework.context.ApplicationEventPublisher;

public class JpaListener {

    @Setter
    private static ApplicationEventPublisher publisher;

    @PrePersist
    public void prePersist(Object entity) {
        if (publisher == null) return;

        publisher.publishEvent(new PrePersistEvent<>(entity));
    }

    @PostPersist
    public void postPersist(Object entity) {
        if (publisher == null) return;

        publisher.publishEvent(new PostPersistEvent<>(entity));
    }

    @PreUpdate
    public void preUpdate(Object entity) {
        if (publisher == null) return;

        publisher.publishEvent(new PreUpdateEvent<>(entity));
    }

    @PostUpdate
    public void postUpdate(Object entity) {
        if (publisher == null) return;

        publisher.publishEvent(new PostUpdateEvent<>(entity));
    }

    @PreRemove
    public void preRemove(Object entity) {
        if (publisher == null) return;

        publisher.publishEvent(new PreRemoveEvent<>(entity));
    }

    @PostRemove
    public void postRemove(Object entity) {
        if (publisher == null) return;

        publisher.publishEvent(new PostRemoveEvent<>(entity));
    }
}

File: api/src/main/kotlin/net/blueshell/api/base/JpaListener.kt
package net.blueshell.api.base

import jakarta.persistence.*
import net.blueshell.api.common.event.jpa.*
import org.springframework.context.ApplicationEventPublisher

class JpaListener {
    @PrePersist
    fun prePersist(entity: Any?) {
        if (publisher == null) return

        publisher.publishEvent(PrePersistEvent<Any?>(entity))
    }

    @PostPersist
    fun postPersist(entity: Any?) {
        if (publisher == null) return

        publisher.publishEvent(PostPersistEvent<Any?>(entity))
    }

    @PreUpdate
    fun preUpdate(entity: Any?) {
        if (publisher == null) return

        publisher.publishEvent(PreUpdateEvent<Any?>(entity))
    }

    @PostUpdate
    fun postUpdate(entity: Any?) {
        if (publisher == null) return

        publisher.publishEvent(PostUpdateEvent<Any?>(entity))
    }

    @PreRemove
    fun preRemove(entity: Any?) {
        if (publisher == null) return

        publisher.publishEvent(PreRemoveEvent<Any?>(entity))
    }

    @PostRemove
    fun postRemove(entity: Any?) {
        if (publisher == null) return

        publisher.publishEvent(PostRemoveEvent<Any?>(entity))
    }

    companion object {
        private var publisher: ApplicationEventPublisher? = null

        @JvmStatic
        fun setPublisher(eventPublisher: ApplicationEventPublisher?) {
            publisher = eventPublisher
        }
    }
}
